### ISSUE TITLE
Mac OS compilation issue...

### ISSUE THREAD

**Issue by patniemeyer** [alicevision/AliceVision#444]
Attempting to build on Mac OS 10.13 I get:

`CMake Error at src/CMakeLists.txt:186 (message):
  The compiler does not support the CXX11 standard.`

**Comment by fabiencastan** [comment]
What compiler version are you using?

**Comment by patniemeyer** [comment]
I have XCode 9.4 installed.  clang and  gcc report the following:
`Apple LLVM version 9.1.0 (clang-902.0.39.2)`
(Do I need XCode 9.4.1?)

Unrelated question: will AliceVision / Meshroom be useful on a Mac that does not have an NVidia card? 

thanks.

**Comment by fabiencastan** [comment]
CUDA is required for depth map computation.
So without CUDA you will be able to compute the sparse reconstruction from structure from motion (point cloud and cameras stored in Alembic file format) but you will not be able to create a dense mesh.

@simogasp Do you know the compiler versions required for C++11 on macos?

**Comment by simogasp** [comment]
@patniemeyer can u please post the log of the messages, including the command line you use to do cmake?

**Comment by patniemeyer** [comment]
Sure - I'm just following the build instructions for Mac in INSTALL.md.

`pat@desk:~/Desktop/alicevision/build  $ cmake -DCMAKE_BUILD_TYPE=Release -G "Xcode" ../AliceVision
-- ----------------------------------------
-- ALICEVISION_BUILD_DEPENDENCIES: OFF
-- ALICEVISION_BUILD_TESTS: OFF
-- AV_BUILD_JPEG: ON
-- AV_BUILD_PNG: ON
-- CMAKE_BUILD_TYPE: Release
-- DEPS_CMAKE_BUILD_TYPE: Release
-- CMAKE_INSTALL_PREFIX: /usr/local
-- CMAKE_INSTALL_RPATH: @loader_path/../lib:@loader_path
-- ----------------------------------------
CMake version: 3.11
-- ** ALICEVISION_BUILD_DOC: 'AUTO'
-- ** ALICEVISION_USE_OPENMP: 'ON'
-- ** ALICEVISION_USE_CCTAG: 'AUTO'
-- ** ALICEVISION_USE_POPSIFT: 'AUTO'
-- ** ALICEVISION_USE_OPENGV: 'AUTO'
-- ** ALICEVISION_USE_ALEMBIC: 'AUTO'
-- ** ALICEVISION_USE_UNCERTAINTYTE: 'AUTO'
-- ** ALICEVISION_USE_CUDA: 'ON'
-- ** ALICEVISION_USE_OPENCV: 'OFF'
-- Detected Host CPU: skylake-avx512
-- Checking for C++11 compiler
-- Checking for C++11 compiler - unavailable
CMake Error at src/CMakeLists.txt:186 (message):
  The compiler does not support the CXX11 standard.


-- Configuring incomplete, errors occurred!
See also "/Users/pat/Desktop/alicevision/build/CMakeFiles/CMakeOutput.log".
pat@desk:~/Desktop/alicevision/build  $ clang --version
Apple LLVM version 9.1.0 (clang-902.0.39.2)
Target: x86_64-apple-darwin17.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
`

**Comment by simogasp** [comment]
can you start from scratch, ie remove the build directory and redo the whole thing. There are some missing lines at the beginning, probably because you already launched once.
The very first 2 lines should read something like
```
-- The C compiler identification is AppleClang 8.0.0.8000042
-- The CXX compiler identification is AppleClang 8.0.0.8000042
```
(but with your version of Apple clang)
Also, can you check the ouptut of`cc --version`, `c++ --version`,`gcc --version`, `g++ --version`

**Comment by patniemeyer** [comment]
Sure, here is a clean attempt:

`pat@desk:~/Desktop/alicevision  $ mkdir build && cd build
pat@desk:~/Desktop/alicevision/build  $ cmake -DCMAKE_BUILD_TYPE=Release -G "Xcode" ../AliceVision
-- The C compiler identification is AppleClang 9.1.0.9020039
-- The CXX compiler identification is AppleClang 9.1.0.9020039
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- ----------------------------------------
-- ALICEVISION_BUILD_DEPENDENCIES: OFF
-- ALICEVISION_BUILD_TESTS: OFF
-- AV_BUILD_JPEG: ON
-- AV_BUILD_PNG: ON
-- CMAKE_BUILD_TYPE: Release
-- DEPS_CMAKE_BUILD_TYPE: Release
-- CMAKE_INSTALL_PREFIX: /usr/local
-- CMAKE_INSTALL_RPATH: @loader_path/../lib:@loader_path
-- ----------------------------------------
CMake version: 3.11
-- ** ALICEVISION_BUILD_DOC: 'AUTO'
-- ** ALICEVISION_USE_OPENMP: 'ON'
-- ** ALICEVISION_USE_CCTAG: 'AUTO'
-- ** ALICEVISION_USE_POPSIFT: 'AUTO'
-- ** ALICEVISION_USE_OPENGV: 'AUTO'
-- ** ALICEVISION_USE_ALEMBIC: 'AUTO'
-- ** ALICEVISION_USE_UNCERTAINTYTE: 'AUTO'
-- ** ALICEVISION_USE_CUDA: 'ON'
-- ** ALICEVISION_USE_OPENCV: 'OFF'
-- target changed from "" to "auto"
-- Detected Host CPU: skylake-avx512
-- Performing Test check_c_compiler_flag__march_skylake_avx512
... # I elided some tests check output here...
-- Performing Test check_cxx_compiler_flag__mno_avx512vbmi - Success
-- Checking for C++11 compiler
-- Checking for C++11 compiler - unavailable
CMake Error at src/CMakeLists.txt:186 (message):
  The compiler does not support the CXX11 standard.

-- Configuring incomplete, errors occurred!
See also "/Users/pat/Desktop/alicevision/build/CMakeFiles/CMakeOutput.log".`

and here are the compiler versions:

`+ cc --version
Apple LLVM version 9.1.0 (clang-902.0.39.2)
Target: x86_64-apple-darwin17.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
+ c++ --version
Apple LLVM version 9.1.0 (clang-902.0.39.2)
Target: x86_64-apple-darwin17.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
+ gcc --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk/usr/include/c++/4.2.1
Apple LLVM version 9.1.0 (clang-902.0.39.2)
Target: x86_64-apple-darwin17.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
+ g++ --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk/usr/include/c++/4.2.1
Apple LLVM version 9.1.0 (clang-902.0.39.2)
Target: x86_64-apple-darwin17.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin`

**Comment by simogasp** [comment]
That's really strange, I thought there was some kind of configuration error but it is not.
It is one the latest version of apple clang and it must support c++11 even by default...

can you share the result of
```
grep -A55 -B4 c++11 CMakeFiles/CMakeOutput.log
```
(from your build directory)

**Comment by simogasp** [comment]
also in an unrelated issue, I don't think you'll be able to compile the cuda part (MVS), as **CUDA 9.2** supports **xcode 9.2** (https://docs.nvidia.com/cuda/cuda-installation-guide-mac-os-x/index.html#system-requirements) and according to your clang version you have **xcode 9.4** (https://gist.github.com/yamaya/2924292#file-xcode-clang-vers-L244). (and I assuming that you have cuda 9.2, otherwise it will be even worst :-( ).
But maybe it will work, i don't know. Just be aware of this possible issue. The usual workaround is to install an older version of xcode supported by cuda.

**Comment by flyandi** [comment]
The default gcc on apple is outdated. In order to enable C++11 support you would need to install gcc manually via homebrew and add enable explicit C++11 in the CMakeLists file.

```brew install gcc```

then modify the CMakeLists and add the following lines below the `project` line: 

```
set(CMAKE_CXX_STANDARD 11)
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
```

This will enable C++11 support.

**Comment by simogasp** [comment]
uhm it's stange because its gcc is a link to the apple compiler.
Instead, I just realized that the check for c++11 is done twice. Can you try to comment out the section
https://github.com/alicevision/AliceVision/blob/develop/src/CMakeLists.txt#L182
till line 189 and redo everything from scratch. 
@fabiencastan I think that section can be removed as https://github.com/alicevision/AliceVision/blob/develop/src/CMakeLists.txt#L31 should be more than enough to enforce C++11. (Why it fails, that's another interesting question...)

**Comment by flyandi** [comment]
Also on another note, you might still have compilation issues since certain libraries don't auto-resolve. You need to install the following libraries:

```
brew install eigen 
brew install ceres-solver
```

After that you need to install opengv:
```
git clone https://github.com/laurentkneip/opengv
cd opengv
mkdir build & cd build
cmake ../opengv 
make
make install
```

Wait until that is completed and then run the AliceVision build command.

**Comment by flyandi** [comment]
@simogasp yes that worked as well.

However now the `uncertaintyTE` is not found :-D

**Comment by simogasp** [comment]
cool!
As for uncertaintyTE it's ok, it is not a mandatory dependency, you should be able to proceed and compile fine.

**Comment by flyandi** [comment]
Nvm. it failed on `geogram` now.

**Comment by flyandi** [comment]
It's not installed or available on my system.

**Comment by simogasp** [comment]
ok
BTW, you can install the dependencies wherever your want in ur system, you just than need to pass the proper `-DDEPENDENCY_DIR:PATH=/path/to/install` at cmake time.
see https://github.com/alicevision/AliceVision/blob/develop/INSTALL.md#cmake-options

Or, this take longer, you can build all the needed dependencies with alicevision with https://github.com/alicevision/AliceVision/blob/develop/INSTALL.md#building-the-project-with-embedded-dependencies

**Comment by flyandi** [comment]
Ok let me try this.

**Comment by flyandi** [comment]
Yeah no go on the embedded.. too many errors like `installation or configuration problem: assembler cannot create object files`.. etc.

**Comment by flyandi** [comment]
Well I am stuck again on 
```
make[5]: *** No rule to make target `/code/repo/a/build/external/png_build/lib/libpng.so', needed by `src/libOpenImageIO/libOpenImageIO.1.8.11.dylib'.  Stop.
```

**Comment by simogasp** [comment]
did libpng build correctly? If you
```
make png
```
does it build?

**Comment by flyandi** [comment]
Yep but for some reason it does not create the libpng.so ..

**Comment by fabiencastan** [comment]
Maybe created in a different folder on macos, like lib64 instead of lib or something like that.

**Comment by simogasp** [comment]
because it's a OSX and it must create a libpng.dylib. Is it there in `/code/repo/a/build/external/png_build/lib/`?
Otherwise try to remove the whole `/code/repo/a/build/external/openimageio`

**Comment by flyandi** [comment]
I am running it again right now.

**Comment by flyandi** [comment]
k .. i have the same result now:

libpng.a
libpng.dylib
libpng.la
libpng16.16.dylib
libpng16.a
libpng16.dylib
libpng16.la
pkgconfig

However it complaints that it needs the .so ...

**Comment by fabiencastan** [comment]
Yes, I have never tried the all-in-one build on macos...

On this line you have an hard-coded relative path to "libpng.so":
https://github.com/alicevision/AliceVision/blob/develop/CMakeLists.txt#L214
You need to replace "libpng.so" by "libpng.dylib" if MACOS.

**Comment by simogasp** [comment]
Argggg!
So basically everywhere in that file.. You need to replace all the .so in CMakeLists.
But it's not the right thing to do, we need to fix this then

**Comment by flyandi** [comment]
Yeah I just noticed that as well :-D 

I changed those and running it again.

**Comment by flyandi** [comment]
So I got all the way to this:

-- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE)
-- Could NOT find Ghostscript (missing: GHOSTSCRIPT_EXECUTABLE)
-- Could NOT find GLPK (missing: GLPK_LIBRARY GLPK_INCLUDE_DIR GLPK_PROPER_VERSION_FOUND) (Required is at least version "4.33")
-- Could NOT find ILOG (missing: ILOG_CPLEX_LIBRARY ILOG_CPLEX_INCLUDE_DIR)
-- Could NOT find COIN (missing: COIN_INCLUDE_DIR COIN_CBC_LIBRARY COIN_CBC_SOLVER_LIBRARY COIN_CGL_LIBRARY COIN_CLP_LIBRARY COIN_COIN_UTILS_LIBRARY COIN_OSI_LIBRARY COIN_OSI_CBC_LIBRARY COIN_OSI_CLP_LIBRARY)
-- Could NOT find SOPLEX (missing: SOPLEX_LIBRARY SOPLEX_INCLUDE_DIR)
-- Could NOT find OpenMP_C (missing: OpenMP_C_FLAGS OpenMP_C_LIB_NAMES)
-- Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES)
-- Could NOT find OpenMP (missing: OpenMP_C_FOUND OpenMP_CXX_FOUND)

**Comment by flyandi** [comment]
Oh ok .. I see the error .. brewing `cuda` right now via `brew install Caskroom/cask/cuda`

**Comment by flyandi** [comment]
However it fails now at `MeshDenoiser` 

```clang: error: unsupported argument 'libomp' to option 'fopenmp='```

**Comment by simogasp** [comment]
yes because the AppleClang does not support openmp.
@fabiencastan Is openmp mandatory for MeshDenoiser?

**Comment by fabiencastan** [comment]
No. It is not supposed to add openmp flags if it is not detected:
https://github.com/alicevision/MeshSDFilter/blob/b7dfeed64be90f2eff49345cf65451b700d3a417/CMakeLists.txt#L32

**Comment by fabiencastan** [comment]
But openmp is enabled in your configuration:
```
** Enable OpenMP parallelization: 1
```

**Comment by fabiencastan** [comment]
Is there no way to get openmp support on osx? The performance will suffer a lot by disabling openmp.

**Comment by simogasp** [comment]
it;'s not officially supported on AppleClang, there are some workarounds
https://iscinumpy.gitlab.io/post/omp-on-high-sierra/

Otherwise, you need to install clang with homebrew as I do and use that one to build. On the other hand, that it is likely to fail building the cuda part as it is not supported on OSX. 
Depending on your version of CUDA, you may try to install clang 4 or 5 but it is not guaranteed.

**Comment by flyandi** [comment]
Well u won't believe it but I just got the line ```** BUILD SUCCEEDED **```

So what I did was disabling OpenMP (and actually removing the sections from the dependency CMake files) and also turning CUDA off.

**Comment by flyandi** [comment]
Ok so now that I have both meshroom and the library working, I am getting stuck on ```No valid initial pair found automatically``` .. 

is there some sample image I can test the library + meshroom before I run my own images?

**Comment by simogasp** [comment]
You can try one of these datasets (maybe `Fountain`)
https://github.com/alicevision/SfM_quality_evaluation

**Comment by flyandi** [comment]
Well - it went through till it had to calculate the depth map .. 

```aliceVision_depthMapEstimation: command not found```

It looks like it was never compiled. I probably bet that is the one that needs CUDA, right?

**Comment by simogasp** [comment]
Yes, without cuda you cannot use the second part of the pipeline for dense reconstruction and meshing (MVS)

**Comment by flyandi** [comment]
So I been trying to use XCODE 9.2.0 which contains the 9.0.0 LLVM and still no luck. Now half of the math.c library is missing .. argh! So how did you guys actually got it compiled under High Sierra? (10.13)?

**Comment by flyandi** [comment]
Btw this is the problem now:

```Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cmath(324): error: the global scope has no "isunordered"```

+ another 30 like these with different methods. Looking at cmath.h it seems everything is there:

![image](https://user-images.githubusercontent.com/329973/43020724-d185c43a-8c15-11e8-9ca0-47c5cbe3a5ba.png)

**Comment by flyandi** [comment]
Here is the full log: https://pastebin.com/X32KhQYQ

**Comment by fabiencastan** [comment]
That's strange. Maybe related to a problem with the cuda runtime installation.
Similar error message here: https://cmake.org/pipermail/cmake/2010-March/036105.html

How did you install cuda on this machine?
Did you already compiled something using cuda on it?

**Comment by dlazares** [comment]
@flyandi what lines did you change to get rid of the OpenMP error? I'm stuck at that point right now

**Comment by ryanfb** [comment]
Thanks partially to the advice and help in this thread, I've just finished writing up a [blog post](https://ryanfb.github.io/etc/2018/08/17/alicevision_and_meshroom_on_mac_os_x.html) and [associated Homebrew tap](https://github.com/ryanfb/homebrew-alicevision) for installing AliceVision on Mac OS X.

**Comment by NextDesign1** [comment]
I was able to compile without any issues using clang 6.0.0 (which includes OpenMP support) and all dependent libraries through macports. The only trick is that you have to be careful to tell CMake to use that compiler instead of the default AppleClang compiler by setting CMAKE_CXX_COMPILER and CMAKE_C_COMPILER to /opt/local/bin/clang++ and /opt/local/bin/clang, respectively.

**Comment by fabiencastan** [comment]
So build is now possible thanks to @ryanfb.
Next step is to provide binaries: https://github.com/alicevision/meshroom/issues/204

**Comment by FunWithFlaggs** [comment]
which nvcc's are supported? i finally got ~80% through the build but it failed with:
`nvcc fatal   : The version ('10.0') of the host compiler ('Apple clang') is not supported`

**Comment by simogasp** [comment]
it depends on your cuda version. For example the latest cuda 10 (https://docs.nvidia.com/cuda/cuda-installation-guide-mac-os-x/index.html) supports apple LLVM 9.0.0 that correspond to the one shipped with XCode 9.0-9.2 (see here https://gist.github.com/yamaya/2924292)

If your xcode version is too newer the only thing to do i'm afraid is to downgrade xcode installing an older version.
